### PR TITLE
Node input validation and help text

### PR DESCRIPTION
**Pull request by SmittieC**
## Description

1. Added help text to some fields. These are populated from the field descriptions.
2. Input validation for nodes. Validation happens on the BE and is propegated to the UI on save or load

## User Impact

Users will see when their input is wrong, but will not be prevented from saving the pipeline.

### Demo

demo

### Docs

N/A

### CODE REVIEW

File: apps/pipelines/models.py [hunk 1]
```diff
@@ -95,16 +95,32 @@ def set_nodes(self, nodes: list[FlowNode]) -> None:
...
+    def validate(self) -> dict:
+        """Validate the pipeline nodes and return a dictionary of errors"""
+        from apps.pipelines.nodes import nodes as pipeline_nodes
+
+        errors = {}
+
+        for node in self.node_set.all():
+            node_class = getattr(pipeline_nodes, node.type)
+            try:
+                node_class(**node.params)
```
Comment (snopoke) [inline apps/pipelines/models.py:107]: nit: use 'model_validate' to make this clearer
```suggestion
                node_class.model_validate(node.params)
```

File: apps/pipelines/nodes/nodes.py [hunk 2]
```diff
@@ -201,9 +207,18 @@ def _get_current_datetime(self, session):
...
-    recipient_list: str
+    recipient_list: str = Field(description="A comma-separated list of email addresses")
     subject: str
 
+    @field_validator("recipient_list", mode="before")
+    def recipient_list_has_valid_emails(cls, value):
+        email_pattern = re.compile(r"^[\w\.-]+@[\w\.-]+\.\w+$")
+        emails = [email.strip() for email in value.split(",")]
+        for email in emails:
+            if not email_pattern.match(email):
```
Comment (snopoke) [inline apps/pipelines/nodes/nodes.py:218]: Django already has an email validator: `django.core.validators.validate_email`

File: apps/pipelines/nodes/nodes.py [hunk 3]
```diff
@@ -394,16 +411,36 @@ def _create_model_from_data(value_data, model_name: str):
...
+        return value
+
+
+class ExtractStructuredData(ExtractStructuredDataNodeMixin, LLMResponse, StructuredDataSchemaValidatorMixin):
     __human_name__ = "Extract Structured Data"
     __node_description__ = "Extract structured data from the input"
-    data_schema: ExpandableText
+    data_schema: ExpandableText = Field(
+        default='{"name": "the name of the user"}',
+        description="A key-value pair where the key is the name of the field and the value the description",
```
Comment (snopoke) [inline apps/pipelines/nodes/nodes.py:433]: This doesn't have to be just key-value pair:
```suggestion
        description="A JSON data structure with descriptions for each object key.",
```

File: apps/pipelines/nodes/nodes.py [hunk 4]
```diff
@@ -394,16 +411,36 @@ def _create_model_from_data(value_data, model_name: str):
...
 
 
-class ExtractParticipantData(ExtractStructuredDataNodeMixin, LLMResponse):
+class ExtractParticipantData(ExtractStructuredDataNodeMixin, LLMResponse, StructuredDataSchemaValidatorMixin):
     __human_name__ = "Extract Participant Data"
     __node_description__ = "Extract structured data and saves it as participant data"
-    data_schema: ExpandableText
+    data_schema: ExpandableText = Field(
+        default='{"name": "the name of the user"}',
+        description="A key-value pair where the key is the name of the field and the value the description",
```
Comment (snopoke) [inline apps/pipelines/nodes/nodes.py:442]:
```suggestion
        description="A JSON data structure with descriptions for each object key.",
```

File: assets/javascript/apps/pipeline/stores/pipelineManagerStore.ts [hunk 5]
```diff
@@ -12,11 +12,14 @@ const usePipelineManagerStore = create<PipelineManagerStoreType>((set, get) => (
   dirty: false,
   isSaving: false,
   isLoading: true,
+  errors: {},
   loadPipeline: async (pipelineId: number) => {
     set({isLoading: true});
     apiClient.getPipeline(pipelineId).then((pipeline) => {
       if (pipeline) {
         set({currentPipeline: pipeline, currentPipelineId: pipelineId});
+        // TODO: Remove this below and read erros from the ReactFlowJsonObject
```
Comment (snopoke) [inline assets/javascript/apps/pipeline/stores/pipelineManagerStore.ts:21]: TODO?

File: assets/javascript/apps/pipeline/widgets.tsx [hunk 6]
```diff
@@ -317,13 +323,17 @@ export function HistoryTypeWidget({
...
-export function InputField({label, children}: React.PropsWithChildren<{ label: string | ReactNode }>) {
+export function InputField({label, help_text, inputError, children}: React.PropsWithChildren<{ label: string | ReactNode, help_text: string, inputError?: string | undefined }>) {
   return (
     <>
       <div className="form-control w-full capitalize">
         <label className="label font-bold">{label}</label>
         {children}
       </div>
+      <div className="flex flex-col">
+        {<small className="text-red-500">{inputError}</small>}
```
Comment (snopoke) [inline assets/javascript/apps/pipeline/widgets.tsx:334]: Why the extra `{}`?

**Review by snopoke — COMMENTED** [review]
Small comments but otherwise looks good

File: apps/pipelines/nodes/nodes.py (hunk 2 above)
Comment (SmittieC) [reply apps/pipelines/nodes/nodes.py]: Nice, here it is: https://github.com/dimagi/open-chat-studio/pull/857/commits/a8631e6f0c0c070ddfa018e2b0c58bb6e1ee660a

File: assets/javascript/apps/pipeline/stores/pipelineManagerStore.ts (hunk 5 above)
Comment (SmittieC) [reply assets/javascript/apps/pipeline/stores/pipelineManagerStore.ts:21]: Oh yeah that is not relevant anymore. I'm keeping the errors on this store level. I'll remove this

File: assets/javascript/apps/pipeline/widgets.tsx (hunk 6 above)
Comment (SmittieC) [reply assets/javascript/apps/pipeline/widgets.tsx:334]: Good question. Not sure. I'll remove

File: apps/pipelines/nodes/nodes.py (hunk 3 above)
Comment (SmittieC) [reply apps/pipelines/nodes/nodes.py:435]: Technically it does have to be a dict-like object. We don't support lists at the top-level currently. I can't remember why exactly, but thinking about it, what would that actually mean and what schema do we give the LLM?

Comment (snopoke) [reply apps/pipelines/nodes/nodes.py:435]: That's fine, maybe we can say a `JSON object structure`?

**Review by snopoke — APPROVED** [review]

File: apps/pipelines/nodes/nodes.py (hunk 3 above)
Comment (SmittieC) [reply apps/pipelines/nodes/nodes.py:435]: Sure, I'll update. I see you've approve, so Ima merge once I updated. Thanks!

Comment (SmittieC) [reply apps/pipelines/nodes/nodes.py]: https://github.com/dimagi/open-chat-studio/pull/857/commits/e828ec811ea4cded658edecabb4c1480417af5a9

**Pull request merged**